src/core/tranche/distributor.sol

```diff
@@ -0,0 +1,198 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+import "ds-note/note.sol";
+
+contract PileLike {
+    function want() public returns (int);
+}
+
+contract QuantLike {
+    uint public debt;
+}
+
+contract OperatorLike {
+
+    function borrow(address, uint) public;
+    function repay(address, uint) public;
+    function balance() public returns (uint);
+    function file(bytes32, bool) public;
+
+    bool public supplyActive;
+    bool public redeemActive;
+
+    QuantLike public quant;
+}
+
+contract Distributor is DSNote {
+
+    // --- Tranches ---
+
+    struct Tranche {
+        bool senior;
+        uint ratio;
+        OperatorLike operator;
+    }
+
+    Tranche[] tranches;
+
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) public auth note { wards[usr] = 1; }
+    function deny(address usr) public auth note { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    PileLike public pile;
+
+    constructor (address pile_) public {
+        wards[msg.sender] = 1;
+        pile = PileLike(pile_);
+    }
+
+    function depend (bytes32 what, address addr) public auth {
+        if (what == "pile") { pile = PileLike(addr); }
+        else revert();
+    }
+
+    // TIN tranche should always be added first
+    function addTranche(uint ratio, address operator_) public auth {
+        Tranche memory t;
+        t.ratio = ratio;
+        t.operator = OperatorLike(operator_);
+        tranches.push(t);
+    }
+
+    function ratioOf(uint i) public auth returns (uint) {
+        return tranches[i].ratio;
+    }
+
+    // if capital should flow through, all funds in reserve should be moved in pile
+    function tradFlow(bool poolClosing) public auth {
+        if (!poolClosing) {
+            for (uint i = 0; i < tranches.length; i++) {
+                // calculates how much money is in the reserve, transfers all of this balance to the pile
+                uint wadR = tranches[i].operator.balance();
+                tranches[i].operator.borrow(address(pile), uint(wadR));
+            }
+        } else {
+            // payout
+            waterfallRepay();
+        }
+    }
+
+    function customFlow(bool poolClosing) public auth {
+        if (!poolClosing) {
+            poolOpen();
+        } else {
+            // this should take the money from the pile, repay and close the vault
+            waterfallRepay();
+        }
+    }
+
+    function poolOpen() private {
+        int wad = pile.want();
+        // pile wants money
+        if (wad > 0) {
+            pileWants();
+        } else if (wad < 0) {
+            pileHas();
+        }
+    }
+
+    function pileWants() private {
+        int wad = pile.want();
+        uint max = calcMaxTake();
```

Comment:
If it is generic max take shouldn't be `maxTake` called in the loop for every tranche?